### ISSUE TITLE
[FR] Set transaction class on file system

### ISSUE THREAD

**Issue by nicholasjng** [fsspec/filesystem_spec#1416]
I have an application of transactions at hand where I want to extend the functionality of fsspec's builtin `Transaction`. In particular, I want to write my own stack unwind routine, and I want to yield the transaction object on entry into the context manager like so:

```python
from fsspec.transaction import Transaction

class MyTransaction(Transaction):
    def __enter__(self):
        self.start()
        return self  # enables the use of a transaction in the context.

    def foo(self, *args, **kwargs):
         ...  # do whatever.

fs = MyFileSystem()
with fs.transaction as tx:
    ... # write files, etc.
    tx.foo()
```

Since the `AbstractFileSystem.transaction` member is a base `Transaction`, I have to override the `transaction` member and `start_transaction` method on the file system, even though I only want to change the transaction type.

So Q1: Would it be possible to make a file system class aware of its associated transaction type (maybe through a `transaction_type` property, default being `fsspec.transactions.Transaction`) to avoid having to overload these methods?

And Q2: Would it be feasible to `return self` in the base `Transaction`'s context entry, to be able to use the object?

Thanks in advance!

**Comment by martindurant** [comment]
Yes, I think it's totally reasonable that Transaction should be subclassable and assigned to filesystem instances (or classes - but I prefer instances). 

There are three major limitations in the current implementation anyway:
- only open() is intercepted, so other write operations like pipe() are still immediate (see discussion linked above)
- in-transaction files don't show up in list or read operations
- on transaction finish, files are committed sequentially, even though a bulk/async operation may be available.

The possible interaction with simplecache is interesting too - that files could be written locally and then uploaded in bulk when the transaction finishes. This is like a totally different conception of a transaction.

**Comment by nicholasjng** [comment]
Thanks! The cache case sounds interesting. I'm actually interested in conducting versioning operations (think `git tag`s, `git merge`s etc.) in the file list unwinding, so at least I would require sequential consistency there (I am basically treating it as a FIFO stack right now), but I see the potential of async there, too.

I can send in a PR with at least the transaction type part if you're interested. My rough idea is this:

```python
from fsspec.transaction import Transaction

class MyTransaction(Transaction):
    ...

class MyFS(AbstractFileSystem)
    _transaction_type = MyTransaction  # or transaction_type=MyTransaction in __init__ for instance-level

    @property
    def transaction_type() -> type[Transaction]:
        return self._transaction_type

    @property
    def transaction() -> Transaction:
        if self._transaction is None:
            self._transaction = self.transaction_type()
        return self._transaction
```

If you have any objections / design ideas, do let me know!

**Comment by martindurant** [comment]
Closed by #1424
